src/einsteinpy/symbolic/tensor.py

```diff
@@ -1,55 +1,327 @@
+        return obj
 
-    def tensor(self):
+    def __repr__(self):
+        return self._print()
+
+    def __str__(self):
+        return str(self.__repr__())
+
+    def __call__(self, *args, **kwargs):
+        new = IndexedTensor(self, args, **kwargs)
+        return new.doit()
+
+    def covariance_transform(self, *indices):
         """
-        Returns the sympy Array
+        Return the array associated with this tensor with indices set according
+        to arguments.
+
+        Parameters
+        ----------
+        indices : TensorIndex
+            Defines the covariance and contravariance of the returned array.
+
+        Examples
+        --------
+        >>> from sympy import diag, symbols, sin
+        >>> from einsteinpy.symbolic.tensor import indices
+        >>> from einsteinpy.symbolic.metric import Metric
+        >>> t, r, th, ph = symbols('t r theta phi')
+        >>> schwarzschild = diag(1-1/r, -1/(1-1/r), -r**2, -r**2*sin(th)**2)
+        >>> g = Metric('g', [t, r, th, ph], schwarzschild)
+        >>> mu, nu = indices('mu nu', g)
+        >>> g.covariance_transform(mu, nu)
+        [[1/(1 - 1/r), 0, 0, 0], [0, 1 - 1/r, 0, 0], [0, 0, -1/r**2, 0], [0, 0, 0, -1/(r**2*sin(theta)**2)]]
 
-        Returns
-        -------
-        ~sympy.tensor.array.dense_ndim_array.ImmutableDenseNDimArray
-            Sympy Array object
-        
         """
-        return self.arr
+        array = self.as_array()
+        for pos, idx in enumerate(indices):
+            if idx.is_up ^ (self.covar[pos] > 0):
+                if idx.is_up:
+                    metric = idx.tensor_index_type.metric.as_inverse()
+                else:
+                    metric = idx.tensor_index_type.metric.as_array()
+                new = tensorcontraction(tensorproduct(metric, array), (1, 2 + pos))
+                permu = list(range(len(indices)))
+                permu[0], permu[pos] = permu[pos], permu[0]
+                array = permutedims(new, permu)
+        return array
 
     def simplify(self):
         """
-        Returns a simplified Tensor
-
-        Returns
-        -------
-        ~einsteinpy.symbolic.tensor.Tensor
-            Simplified Tensor
+        Replace the stored array associated with this tensor with a simplified
+        version. This method also replaces the entry in ReplacementManager.
 
         """
-        return sympy.simplify(self.tensor())
+        array = simplify(self.as_array())
+        self._array = array
+        ReplacementManager.replace(self, array)
+        return array
+
+
+class Index(TensorIndex):
+    """
+    Class for a symbolic representation of a tensor index with respect to a metric.
+    """
+
+    def __new__(cls, symbol, metric, is_up=True, **kwargs):
+        return super().__new__(cls, symbol, metric, is_up=is_up, **kwargs)
+
+    def __neg__(self):
+        return Index(self.name, self.tensor_index_type, (not self.is_up))
+
+
+def expand_tensor(expr, idxs=None):
```

Comment:
This is the function by which performs the calculations represented by a tensor expression. Whenever you are finished doing algebra symbolically, this is the method you use to get an array equivalent to the said expression.